### DISCUSSION TITLE
How To DIsable Middleware Cache?

### DISCUSSION THREAD

**Discussion by Enkratia** [vercel/next.js#59258] [Help]
### Summary

I need to disable cache for routes like "/profile/:path*" and  "/dashboard/:path*
I have spend almost 3 days to find a solution and found nothing

Middleware:
- can't use "revalidatePath"
- can't use "revalidateTag"
- "x-middleware-cache": "no-cache" - doesn't make any difference
- "cache-control": "no-store" - doesn't make any difference

**How To DIsable Middleware Cache For Pages?**

**Comment by andr3wV** [comment]
Does anyone have an answer to this?

**Reply by Enkratia** [reply]
As a workaround you can invoke server action with usage of `headers()` or `cookies()` inside of particular page, which will make page dynamic. Or if you need such functionality for all/many pages you can create helper component, which will execute the server action on those pages (with help of `usePathaname()`) via `React.useEffect()`. 

I hope it understandable.

**Reply by andr3wV** [reply]
Maybe I'm implementing what you're saying incorrectly, but I am having no luck with that method. Here is what I have:

**NoCache.tsx**:
```
"use client";

import { useEffect } from "react";
import { usePathname } from "next/navigation";

const NoCache = () => {
  const pathname = usePathname();

  useEffect(() => {
    if (pathname.startsWith("/dashboard")) {
      // Force the browser to not cache the route
      const headers = new Headers();
      headers.set("Cache-Control", "no-store");
    }
  }, [pathname]);

  return null;
};

export default NoCache;
```

**Layout.tsx** which handles all the routes that I don't want to have cache:

```
import "@/app/(app)/globals.css"
import type { Metadata } from "next"
import Header from "./components/header";
import NoCache from "@/components/NoCache";

export default function DashboardPage() {
  return (
    <>
      <NoCache />
      <h1>Dashboard</h1>
      {children}
    </>
  );
}

```

**Reply by andr3wV** [reply]
The only solution I've found that works is when actually doing the routing:

const timestamp = new Date().getTime(); // timestamp trick to prevent caching 
router.push(`/dashboard?ts=${timestamp}`);

But this is a pretty bad solution. @Enkratia any thoughts?

**Reply by Enkratia** [reply]
@andr3wV

CommonHelper component:

```
"use client";

import React from "react";
import { usePathname } from "next/navigation";

import { checkHeadersAction } from "../../utils/actions";

export const CommonHelper: React.FC = () => {
  const pathname = usePathname();

  React.useEffect(() => {
    const checkHeaders = async () => {
      try {
        await checkHeadersAction();
      } catch (error) {
        console.warn("Failed to get headers");
      }
    };

    checkHeaders();
  }, [pathname]);

  return <></>;
};

```
server action:

```
"use server";

import { headers } from "next/headers";

export const checkHeadersAction = async () => {
  return headers().keys();
};
```

It should make every page dynamic i guess, this is what i did. 
Hope it understandable now.

**Reply by Enkratia** [reply]
@andr3wV, give me know if it works for you

If you need it for dashboard only, you can replace:

`checkHeaders();`

with:

```
if (pathname.startsWith("/dashboard") {
  checkHeaders();
}
```

**Reply by andr3wV** [reply]
Still can't seem to get it to work. Do you put CommonHelper in middleware or in a the layout.tsx for the dashboard route?

**Reply by andr3wV** [reply]
I can't believe this still requires these types of workarounds, it's crazy

**Reply by Enkratia** [reply]
> Still can't seem to get it to work. Do you put CommonHelper in middleware or in a the layout.tsx for the dashboard route?

Layout.tsx, of course, how would you add component to middleware.ts). Root layout of the app (close to body tag)

**Reply by Enkratia** [reply]
But you can add `CommonHelper` to the dashboard layout aswell

**Reply by andr3wV** [reply]
Ya idk it's refusing to work😭

**Reply by andr3wV** [reply]
This is what I have:

**actions.ts**
```
"use server";

import { headers } from "next/headers";

export const checkHeadersAction = async () => {
  return headers().keys();
};
```

**CommonHelper.tsx**
```
"use client";

import React from "react";
import { usePathname } from "next/navigation";
import { checkHeadersAction } from "./actions";

export const CommonHelper: React.FC = () => {
  const pathname = usePathname();

  React.useEffect(() => {
    const checkModalHeader = async () => {
      try {
        await checkHeadersAction();
      } catch (error) {
        console.warn("Failed to get headers");
      }
    };

    checkModalHeader();
  }, [pathname]);

  return null;
};
```

**layout.tsx**
```
import "@/app/(app)/globals.css"
import type { Metadata } from "next"
import Header from "./components/header";
import { CommonHelper } from "./CommonHelper";

export const metadata: Metadata = {
  title: "Dashboard",
}

export default async function Layout({ children }:  {children: React.ReactNode; }) {
    return (
        <>
            <CommonHelper/>
            <Header />
            {children}
        </>
    )
}

```

**Reply by andr3wV** [reply]
> Maybe I'm implementing what you're saying incorrectly, but I am having no luck with that method. Here is what I have:
> 
> **NoCache.tsx**:
> 
> ```
> "use client";
> 
> import { useEffect } from "react";
> import { usePathname } from "next/navigation";
> 
> const NoCache = () => {
>   const pathname = usePathname();
> 
>   useEffect(() => {
>     if (pathname.startsWith("/dashboard")) {
>       // Force the browser to not cache the route
>       const headers = new Headers();
>       headers.set("Cache-Control", "no-store");
>     }
>   }, [pathname]);
> 
>   return null;
> };
> 
> export default NoCache;
> ```
> 
> **Layout.tsx** which handles all the routes that I don't want to have cache:
> 
> ```
> import "@/app/(app)/globals.css"
> import type { Metadata } from "next"
> import Header from "./components/header";
> import NoCache from "@/components/NoCache";
> 
> export default function DashboardPage() {
>   return (
>     <>
>       <NoCache />
>       <h1>Dashboard</h1>
>       {children}
>     </>
>   );
> }
> ```

Update: This actually does work. I thought it wasn't working because my `<Link>` components had prefetch enabled. @Enkratia thank you for you're help, I really appreciate it!

**Comment by Enkratia** [comment]
@andr3wV, glad to know, thank you for reply :)

**Reply by Enkratia** [reply]
@andr3wV, you shouldn't disable prefetch to get desired result. Adding these 2 small files in the project is all what you need to disable middleware cache, i just rechecked it myself.

(This is important to clarify that this topic is about middleware cache, not any cache.)
